test/matrix_test.go

```diff
@@ -0,0 +1,561 @@
+	defer cancel()
+	c, namespace := setup(ctx, t, requireAlphaFeatureFlag)
+	knativetest.CleanupOnInterrupt(func() { tearDown(ctx, t, c, namespace) }, t.Logf)
+	defer tearDown(ctx, t, c, namespace)
+	t.Logf("Creating Task in namespace %s", namespace)
+	task := parse.MustParseV1Task(t, fmt.Sprintf(`
+metadata:
+  name: mytask
+  namespace: %s
+spec:
+  params:
+    - name: exit-code
+  steps:
+    - name: echo
+      image: alpine
+      script: |
+        exit "$(params.exit-code)"
+`, namespace))
+
+	if _, err := c.V1TaskClient.Create(ctx, task, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Task `%s`: %s", task.Name, err)
+	}
+
+	pipelineRun := parse.MustParseV1PipelineRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  serviceAccountName: "default"
+  pipelineSpec:
+    tasks:
+      - name: exit-with-zero-and-non-zero
+        taskRef:
+          name: mytask
+        matrix:
+          params:
+          - name: exit-code
+            value:
+              - "0"
+              - "0"
+              - "1"
+`, helpers.ObjectNameForTest(t), namespace))
+	prName := pipelineRun.Name
+
+	t.Logf("Creating PipelineRun %s", prName)
+	if _, err := c.V1PipelineRunClient.Create(ctx, pipelineRun, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Pipeline `%s`: %s", prName, err)
+	}
+
+	t.Logf("Waiting for PipelineRun in namespace %s to fail", namespace)
+	if err := WaitForPipelineRunState(ctx, c, prName, timeout, PipelineRunFailed(prName), "PipelineRunFailed", v1Version); err != nil {
+		t.Errorf("Error waiting for PipelineRun to finish: %s", err)
+	}
+
+	pr, err := c.V1PipelineRunClient.Get(ctx, prName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Couldn't get expected PipelineRun %s: %s", pr.Name, err)
+	}
+
+	if pr.Status.GetCondition(apis.ConditionSucceeded).IsTrue() {
+		t.Errorf("Expected PipelineRun to fail but found condition: %s", pr.Status.GetCondition(apis.ConditionSucceeded))
+	}
+
+	taskRun1 := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  params:
+  - name: exit-code
+    value: 0
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: mytask
+    kind: Task
+status:
+  conditions:
+  - status: "True"
+    type: Succeeded
+`, "pr-exit-with-zero-and-non-zero-0", namespace))
+
+	taskRun2 := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  params:
+  - name: exit-code
+    value: 0
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: mytask
+    kind: Task
+status:
+  conditions:
+  - status: "True"
+    type: Succeeded
+`, "pr-exit-with-zero-and-non-zero-1", namespace))
+
+	taskRun3 := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  params:
+  - name: exit-code
+    value: 1
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: mytask
+    kind: Task
+status:
+  conditions:
+  - status: "True"
```

Comment:
I expected this TaskRun to fail 🤔

Reply:
+1; it looks like the tests are only comparing the spec, not the status. can you update to compare the status as well?

Reply:
Added testing for the status